---
title: "Analyzing the sensitivity and specificity of ESVs for discriminating between genomes"
author: "Gaurav Bhatti; Pat Schloss"
date: "09/04/2024"
output:
 github_document:
   html_preview: false
editor_options: 
  chunk_output_type: console
---


```{r, message=FALSE}
library(tidyverse)
library(here)
```


### Determine the number of *rrn* operons across genomes

Our analysis will use full length sequences. But since we know that the sub regions are less diverse than the full length sequence. So does the number of ESVs per genome differ than for full length sequences? Are ESVs as specific when using the V4 region compared to full length sequences?


```{r n_rrn}
count_tibble<- read_tsv(here("data/processed/rrnDB.easv.count_tibble"),
                        col_types = cols(.default=col_character(),
                                         count=col_integer())) |> 
  filter(threshold=="esv") |> 
  select(-threshold)
```

We want to count and plot the number of copies per genome.

```{r}
count_tibble |> 
  filter(region=="v19") |> 
  group_by(genome) |> 
  summarize(n_rrn = sum(count)) |> 
  ggplot(aes(x=n_rrn))+
  geom_histogram(binwidth = 1)

count_tibble |> 
  filter(region=="v19") |> 
  group_by(genome) |> 
  summarize(n_rrn = sum(count)) |> 
  count(n_rrn) |> 
  mutate(fraction= 100*n/sum(n))

```


We see that most genomes (92%) have more than 1 copy of the *rrn* operon. I wonder whether those different copies are the same sequence/ESV...

### Determine the number of ESVs per genome

Considering that most genomes have more than 1 copy of the *rrn* operon, we need to know whether they all have the same ESV. Otherwise we run the risk of splitting a single genome into multiple ESVs.

```{r}
count_tibble |> 
  group_by(region,genome) |> 
  summarise(n_esv= n(),
            n_rrn = sum(count)) |> 
  group_by(region,n_rrn) |> 
  summarize(med_n_esv= median(n_esv),
            mean_n_esv=mean(n_esv),
            lg_n_esv= quantile(n_esv,0.25),
            up_n_esv= quantile(n_esv,0.75)) |> 
  filter(n_rrn==7)


count_tibble |> 
  group_by(region,genome) |> 
  summarise(n_esv= n(),
            n_rrn = sum(count)) |> 
  ggplot(aes(x=n_rrn,y=n_esv,color=region))+
  geom_smooth(method = "lm")+
  geom_point()

```


Surprisingly (or not) the number of unique ESVs increases at the rate of about 2 ESVS per 3 copies of *rrn* operon in the genome.The sub-regions of the 16S rRNA gene have fewer ESVs per *rrn* operon.

### Determine whether an ESV is specific to the genome they are found in.

Instead of looking at the number of ESVs per genome, we want to see the number of genomes per ESV.

```{r}
count_tibble |> 
  group_by(region,easv) |> 
  summarise(n_genomes= n()) |> 
  count(n_genomes) |> 
  mutate(fraction=100*n/sum(n)) |> 
  filter(n_genomes==1)

```

We see that with full length sequences, 82% of the ESVs were unique to the genome. For the sub regions, about 76% of the ESVs were unique to the genome.



### To be determined
* Can we correct for over representation?
* Consider analysis at species, genus, family, etc. levels.
* Consider looking at more broad definition of an ESV (upto 3% differences in sequences).

